src/zarr/abc/codec.py

```diff
@@ -75,13 +77,19 @@ def evolve_from_array_spec(self, array_spec: ArraySpec) -> Self:
         """
         return self
 
-    def validate(self, array_metadata: ArrayMetadata) -> None:
-        """Validates that the codec configuration is compatible with the array metadata.
+    def validate(self, shape: ChunkCoords, dtype: np.dtype[Any], chunk_grid: ChunkGrid) -> None:
```

Comment:
perhaps then we could use keyword-only arguments here

Reply:
I made these keyword-only arguments to give us some flexibility. I did consider keeping `ArrayMetadata` here, but a) since `ArrayMetadata.__init__` calls this method, we run a risk of circularity and b) I think we can categorically state that no codec will ever need access to `ArrayMetadata.attributes`, so that alone argues against using `ArrayMetadata` here as an argument. I think it's much better if we are explicit about only passing the required arguments into this function.